platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/configuration/DaemonJvmToolchainCriteriaOptions.java

```diff
@@ -0,0 +1,103 @@
+import org.gradle.internal.buildoption.StringBuildOption;
+import org.gradle.internal.jvm.inspection.JvmVendor;
+import org.gradle.jvm.toolchain.JavaLanguageVersion;
+import org.gradle.jvm.toolchain.JvmImplementation;
+import org.gradle.jvm.toolchain.internal.DefaultJvmVendorSpec;
+
+import java.util.List;
+
+public class DaemonJvmToolchainCriteriaOptions extends BuildOptionSet<DaemonJvmToolchainSpec> {
+
+    private static List<BuildOption<DaemonJvmToolchainSpec>> options = ImmutableList.of(
+        new ToolchainVersionOption(),
+        new ToolchainVendorOption(),
+        new ToolchainImplementationOption()
+    );
+
+    public static List<BuildOption<DaemonJvmToolchainSpec>> get() {
+        return options;
+    }
+
+    @Override
+    public List<BuildOption<DaemonJvmToolchainSpec>> getAllOptions() {
+        return options;
+    }
+
+    public static class ToolchainVersionOption extends StringBuildOption<DaemonJvmToolchainSpec> {
+
+        public ToolchainVersionOption() {
+            super(BuildPropertiesDefaults.TOOLCHAIN_VERSION_PROPERTY, PropertyOrigin.BUILD_PROPERTIES);
+        }
+
+        @Override
+        public void applyTo(String value, DaemonJvmToolchainSpec settings, Origin origin) {
+            try {
+                int integerValue = Integer.parseInt(value);
+                settings.getLanguageVersion().set(JavaLanguageVersion.of(integerValue));
+            } catch (NumberFormatException e) {
+                origin.handleInvalidValue(value, "the value should be an int");
+            } catch (IllegalArgumentException e) {
+                origin.handleInvalidValue(value, "the value should be a positive int");
+            }
+        }
+    }
+
+    public static class ToolchainVendorOption extends EnumBuildOption<JvmVendor.KnownJvmVendor, DaemonJvmToolchainSpec> {
+
+        public ToolchainVendorOption() {
+            super(BuildPropertiesDefaults.TOOLCHAIN_VENDOR_PROPERTY, JvmVendor.KnownJvmVendor.class, JvmVendor.KnownJvmVendor.values(), BuildPropertiesDefaults.TOOLCHAIN_VENDOR_PROPERTY, PropertyOrigin.BUILD_PROPERTIES);
+        }
+
+        @Override
+        public void applyTo(JvmVendor.KnownJvmVendor value, DaemonJvmToolchainSpec settings, Origin origin) {
+            assertJvmToolchainVersion(settings);
+            settings.getVendor().set(DefaultJvmVendorSpec.of(value));
+        }
+    }
+
+    public static class ToolchainImplementationOption extends EnumBuildOption<JvmImplementation, DaemonJvmToolchainSpec> {
+
+        public ToolchainImplementationOption() {
+            super(BuildPropertiesDefaults.TOOLCHAIN_IMPLEMENTATION_PROPERTY, JvmImplementation.class, JvmImplementation.values(), BuildPropertiesDefaults.TOOLCHAIN_IMPLEMENTATION_PROPERTY, PropertyOrigin.BUILD_PROPERTIES);
+        }
+
+        @Override
+        public void applyTo(JvmImplementation value, DaemonJvmToolchainSpec settings, Origin origin) {
+            assertJvmToolchainVersion(settings);
```

Comment:
This seems like the wrong place to be calling this method. 
It seems more appropriate to call it when we are consuming these property values later.

This code seems to rely on the implicit ordering that `ToolchainVersionOption` is called before the others -- and IMO each of these build options should not need to know about each other

Reply:
A part of the existing `UpdateDaemonJvmModifier` that is used to create the build properties we have the parser living on `DaemonJvmToolchainCriteriaOptions` that is responsible to parse the defined properties values. Meaning that right now the only to know about the existence of `daemon.jvm.toolchain.version` is this class. 

I do agree that this depends on the implicit ordering of `ToolchainVersionOption` and will be great  given a `BuildOption` to have some pre-validation before applying any value since in this case  the vendor requires to have defined the version on the file.

Reply:
This still seems wrong to me. I do not think build options should need to know about each other. The architecture of build options are constructed such that each option is independent of each other. I'm not saying that's a _great_ design, but that's how it is and without changing how build options work, we should respect that architecture.

Given that, we should add this check where these values are _consumed_ -- the place that actually necessitates this constraint. Why do we need a version when defining a vendor? There is a logical reason we have this restriction, so we should make this assertion there. 

I am not sure I fully understand this:
```
Meaning that right now the only to know about the existence of daemon.jvm.toolchain.version is this class.
```